ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/MLSdkAsyncHttpResponseHandler.java

```diff
@@ -80,6 +81,23 @@ public void onHeaders(SdkHttpResponse response) {
         SdkHttpFullResponse sdkResponse = (SdkHttpFullResponse) response;
         log.debug("received response headers: " + sdkResponse.headers());
         this.statusCode = sdkResponse.statusCode();
+        if (MapUtils.isEmpty(sdkResponse.headers())) {
+            return;
+        }
+        List<String> errorsInHeader = sdkResponse.headers().get("x-amzn-ErrorType");
+        if (errorsInHeader == null || errorsInHeader.isEmpty()) {
+            return;
+        }
+        boolean containsThrottlingException = errorsInHeader.stream().anyMatch(str -> str.startsWith("ThrottlingException"));
+        if (containsThrottlingException) {
+            actionListener
+                .onFailure(
```

Comment:
I have a big concern here. Although we call actionListener.onFailure in onHeaders, this doesn't prevent the onStream being invoked(I confirmed it using debugger tool). And we'll still subscribe the response and trigger the actionListener.onFailure again at the MLResponseSubscriber. And this can lead to bugs to call actionListener.onFailure twice.

Reply:
Thanks for the callout. I had the same concern but didn't verify it yet yesterday. Already have this fixed in the new commit.